### ISSUE TITLE
Seemingly version 2.7.9 has problems with frozen strings

### ISSUE THREAD

**Issue by kikonen** [petergoldstein/dalli#704]
Appeared after updating from 2.7.8 to 2.7.9 version.

```
ERROR [2018-10-24T05:00:51.777778 #14063-22175620]  #<ActionView::Template::Error: can't modify frozen String>
xxx/shared/bundle/ruby/2.5.0/gems/dalli-2.7.9/lib/active_support/cache/dalli_store.rb:372:in `force_encoding'
xxx/shared/bundle/ruby/2.5.0/gems/dalli-2.7.9/lib/active_support/cache/dalli_store.rb:372:in `expanded_key'
xxx/shared/bundle/ruby/2.5.0/gems/dalli-2.7.9/lib/active_support/cache/dalli_store.rb:361:in `block in expanded_key'
xxx/shared/bundle/ruby/2.5.0/gems/dalli-2.7.9/lib/active_support/cache/dalli_store.rb:361:in `collect'
xxx/shared/bundle/ruby/2.5.0/gems/dalli-2.7.9/lib/active_support/cache/dalli_store.rb:361:in `expanded_key'
xxx/shared/bundle/ruby/2.5.0/gems/dalli-2.7.9/lib/active_support/cache/dalli_store.rb:340:in `namespaced_key'
xxx/shared/bundle/ruby/2.5.0/gems/dalli-2.7.9/lib/active_support/cache/dalli_store.rb:129:in `read'
xxx/shared/bundle/ruby/2.5.0/gems/actionpack-5.2.1/lib/abstract_controller/caching/fragments.rb:114:in `block in read_fragment'
xxx/shared/bundle/ruby/2.5.0/gems/actionpack-5.2.1/lib/abstract_controller/caching/fragments.rb:162:in `block in instrument_fragment_cache'
xxx/shared/bundle/ruby/2.5.0/gems/activesupport-5.2.1/lib/active_support/notifications.rb:168:in `block in instrument'
xxx/shared/bundle/ruby/2.5.0/gems/activesupport-5.2.1/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
xxx/shared/bundle/ruby/2.5.0/gems/activesupport-5.2.1/lib/active_support/notifications.rb:168:in `instrument'
xxx/shared/bundle/ruby/2.5.0/gems/actionpack-5.2.1/lib/abstract_controller/caching/fragments.rb:162:in `instrument_fragment_cache'
xxx/shared/bundle/ruby/2.5.0/gems/actionpack-5.2.1/lib/abstract_controller/caching/fragments.rb:113:in `read_fragment'
xxx/shared/bundle/ruby/2.5.0/gems/actionview-5.2.1/lib/action_view/helpers/cache_helper.rb:248:in `read_fragment_for'
xxx/shared/bundle/ruby/2.5.0/gems/actionview-5.2.1/lib/action_view/helpers/cache_helper.rb:238:in `fragment_for'
xxx/shared/bundle/ruby/2.5.0/gems/actionview-5.2.1/lib/action_view/helpers/cache_helper.rb:169:in `cache'
xxx/releases/20181024080643/app/views/layouts/application.html.erb:48:in `_app_views_layouts_application_html_erb__3856684245160448392_70278594074300'
```
via application.html.erb
```
<% cache("xxx_#{foobar}", expires_in: NN) do %>
  ...
```

**Comment by adamniedzielski** [comment]
We've upgraded from version 2.7.8 to 2.7.9 as well and are experiencing the same issue.

I believe that this regression was introduced by https://github.com/petergoldstein/dalli/pull/694 /cc @schneems 

It seems like in some cases `key` in `key.force_ending('binary')` can be a frozen string. Before this performance improvement `key` was always being duplicated so there was no issue with calling `force_encoding`.

**Comment by schneems** [comment]
Here's a test case with a frozen string https://github.com/petergoldstein/dalli/pull/705

**Comment by schneems** [comment]
It looks like that test is passing, can you give me any more information about how the key is constructed? Do you have an example of an object being passed into the cache that can trigger this failure?

**Comment by adamniedzielski** [comment]
@schneems I submitted #707. I wrote this failing test after reading the implementation so I'm not 100% sure if that's the scenario that triggers the error in our application. The caching code is so deep down that I fear to embark there.

**Comment by kikonen** [comment]
In my  case it was triggered via application.html.erb

via
```
<% cache("xxx_#{foobar}", expires_in: NN) do %>
  ...
```

**Comment by schneems** [comment]
Closed by #708
